### PR TITLE
feat: Modify Column Type

### PR DESCRIPTION
**Pull request by KKould**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
https://github.com/GreptimeTeam/greptimedb/issues/3517
https://github.com/GreptimeTeam/greptime-proto/pull/146

## What's changed and what's your intention?

Support: `ALTER TABLE <table_name> ALTER COLUMN <column_name> TYPE <data_type>;`
ref: https://www.postgresql.org/docs/current/sql-altertable.html

Tips 1: sql parser does not support `Modify Column` temporarily, I have created a PR: https://github.com/sqlparser-rs/sqlparser-rs/pull/1216

Tips 2: I temporarily use my own fork of `greptime-proto`, and then replace it after merging it at https://github.com/GreptimeTeam/greptime-proto/pull/146

```shell
public=> CREATE TABLE test(i INTEGER, j TIMESTAMP TIME INDEX);
OK 0
public=> ALTER TABLE test ALTER COLUMN I TYPE STRING;
OK 0
public=> DESCRIBE t2;
+--------+----------------------+-----+------+---------+---------------+
| Column | Type                 | Key | Null | Default | Semantic Type |
+--------+----------------------+-----+------+---------+---------------+
| i      | String               |     | YES  |         | FIELD         |
| j      | TimestampMillisecond | PRI | NO   |         | TIMESTAMP     |
+--------+----------------------+-----+------+---------+---------------+
```

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

**Comment by WenyXu** [comment]
Nice job🚀
> Tips: sql parser does not support Modify Column temporarily, I have created a PR: https://github.com/sqlparser-rs/sqlparser-rs/pull/1216

BTW, Should we support both dialects(i.e., `ALTER TABLE <table_name> ALTER COLUMN <column_name> TYPE <data_type>;` and `ALTER TABLE <table_name> MODIFY <column_name> INT` )?

**Comment by KKould** [comment]
https://github.com/GreptimeTeam/greptimedb/issues/3517 mentioned that it supports MySQL dialects as much as possible. I think it is good to only support MySQL.

The current PostgreSQL syntax is only for testing

File: tests/cases/standalone/common/alter/modify_col_not_null.sql [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+ALTER TABLE test ALTER COLUMN j TYPE STRING;
+
+INSERT INTO test VALUES (3, "greptime");
+
+SELECT * FROM test;
+
+DESCRIBE test;
+
+DROP TABLE test;
```
Comment (WenyXu) [inline tests/cases/standalone/common/alter/modify_col_not_null.sql:15]: Add an empty line

File: tests/cases/standalone/common/alter/modify_col.sql [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+ALTER TABLE test ALTER COLUMN I TYPE STRING;
+
+SELECT * FROM test;
+
+ALTER TABLE test ALTER COLUMN j TYPE STRING;
+
+DESCRIBE test;
+
+DROP TABLE test;
```
Comment (WenyXu) [inline tests/cases/standalone/common/alter/modify_col.sql:15]: ditto

File: tests/cases/standalone/common/alter/modify_col.sql
Comment (WenyXu) [inline tests/cases/standalone/common/alter/modify_col.sql:1]: Let's add some failed cases for modifying column types.

**Comment by WenyXu** [comment]
Run `make fmt` for formatting the code.

File: src/table/src/metadata.rs [hunk 3]
```diff
@@ -458,6 +459,118 @@ impl TableMeta {
         Ok(meta_builder)
     }
 
+    fn modify_columns(
```
Comment (WenyXu) [inline src/table/src/metadata.rs:462]: non-blocking: Add more tests to cover branches of the `modify_columns`(Maybe in another PR)

File: tests/cases/standalone/common/alter/modify_col.result [hunk 4]
```diff
@@ -1,39 +1,49 @@
...
-SELECT * FROM test;
+ALTER TABLE test ALTER COLUMN id TYPE STRING;
 
-+---+-------------------------+
-| i | j                       |
-+---+-------------------------+
-| 1 | 1970-01-01T00:00:00.001 |
-| 2 | 1970-01-01T00:00:00.002 |
-+---+-------------------------+
+Error: 1004(InvalidArguments), Not allowed to modify index column id from table test
```
Comment (WenyXu) [inline tests/cases/standalone/common/alter/modify_col.result:19]: nit:
```suggestion
Error: 1004(InvalidArguments), Not allowed to modify index column `id` of table test
```

**Review by WenyXu — COMMENTED** [review]
non-blocking: Let's add some fuzz tests for column type modification(Maybe in another PR).

File: src/store-api/src/metadata.rs [hunk 5]
```diff
@@ -615,6 +618,21 @@ impl RegionMetadataBuilder {
         self.column_metadatas
             .retain(|col| !name_set.contains(&col.column_schema.name));
     }
+
+    /// Modifies columns to the metadata if exist.
+    fn modify_columns(&mut self, columns: Vec<ModifyColumn>) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (KKould) [reply src/store-api/src/metadata.rs]: > Should we throw an error if the column does not exist?

`ModifyColumn::validate` this method has been verified

File: tests/cases/standalone/common/alter/modify_col.result (hunk 4 above)
Comment (KKould) [reply tests/cases/standalone/common/alter/modify_col.result]: as you wish

but single quotes are not used elsewhere (like `add column`)

**Comment by evenyag** [comment]
Since altering column types is a big feature, we might need further discussions and implement it carefully. Could you please describe how you implemented this feature in the issue?

We also need to consider the following things:
- Conversion rule
    - Do we allow altering a string column to an integer column?
    - Do we allow converting int16 to int8 or uint16?
- Should we allow altering time index or tag column types? How to support that?
- How it works?
    - Does it rewrite all SST files?
- Query performance impact
- Should other region engines such as metric and file support this?

What's more, I'd suggest implementing this step by step as it also depends on other crates. For example:
- Support altering column type in the region engine first
- Support this in the region server, which also modifies the `greptime-proto` crate
- Support parsing SQL texts to alter the column type

**Comment by KKould** [comment]
> * Conversion rule
>   
>   * Do we allow altering a string column to an integer column?
>   * Do we allow converting int16 to int8 or uint16?
> * Should we allow altering time index or tag column types? How to support that?
> * How it works?
>   
>   * Does it rewrite all SST files?
> * Query performance impact
> * Should other region engines such as metric and file support this?

- Conversion rule
  - Q 1: Do we allow altering a string column to an integer column?
  - A 2: String can theoretically be converted to any type, so it should be possible
  - Q 2: Do we allow converting int16 to int8 or uint16?
  - A 2: I think it should be allowed, specific reference： `ConcreteDataType::can_arrow_type_cast_to` & `VectorOp::cast`

- Q 3: Should we allow altering time index or tag column types? How to support that?
- A 3: `src/table/src/metadata.rs L479` column verification will be performed here, and modification of timesindex and tag is not allowed.
- How it works?
  - Q 4: Does it rewrite all SST files?
  - A 4: It is definitely not a good idea to completely modify the data, especially for time series databases. I implemented it with reference to add/drop column, and checked whether expect and actual are consistent in compat.rs. When the names are consistent but the types are inconsistent, type conversion will be performed based on `IndexOrDefault::Index`. Therefore, ModifyColumn does not modify the data, but temporarily performs type conversion when reading, and changes the type to the modified type when the next compaction is triggered to avoid type conversion (this can also be done when modify column is performed multiple times in a short period of time, No additional overhead will be incurred)
```rust
/// Index in source batch or a default value to fill a column.
#[derive(Debug)]
enum IndexOrDefault {
    /// Index of the column in source batch.
    Index {
        pos: usize,
        cast_type: Option<ConcreteDataType>,
    },
    /// Default value for the column.
    DefaultValue {
        /// Id of the column.
        column_id: ColumnId,
        /// Default value. The vector has only 1 element.
        default_vector: VectorRef,
    },
}
```
- Q 5: Query performance impact？
- A 5: as I said before, the data needs to be type converted before the next compaction, so there is a column type conversion performance problem when reading.
- Q 6: Should other region engines such as metric and file support this?
- A 6: I don't know how to use other engines. hope you can make more suggestions in this regard

**Comment by evenyag** [comment]
> Q 1: Do we allow altering a string column to an integer column?
> A 2: String can theoretically be converted to any type, so it should be possible
> Q 2: Do we allow converting int16 to int8 or uint16?
> A 2: I think it should be allowed, specific reference： ConcreteDataType::can_arrow_type_cast_to & VectorOp::cast

What about error handling?
- We can't cast a string like 'hello' to an integer
- Converting int16 to int8 might overflows
- `arrow::cast` may returns an error

Maybe we can only support a subset of conversion rules that arrow implements.

> src/table/src/metadata.rs L479 column verification will be performed here, and modification of timesindex and tag is not allowed.

Modifying the data type of the time index might be helpful. e.g. Alter a timestamp(second) to timestamp(millisecond). But we can support it later.

For other engines (file, metric), you could refer to their implementation of the `RegionEngine` trait. Maybe we should disallow modifying the column type for those engines.

I'm not going to implement `ALTER COLUMN` in this single PR. We can convert https://github.com/GreptimeTeam/greptimedb/issues/3517 into a tracking issue. You can summarize your design in that issue and list your steps to implement it. Then we can move the discussion to the issue.

It'd be better if we only modify a single crate in each step. e.g.
- Add a new variant for modifying a column to the `AlterRegionRequest` in the store-api
- Support that variant in mito2
- Deal with other region engines
- Add the variant to the gRPC request
- Support it in the region server
- Parser
- Support `ALTER COLUMN` statement

**Pull request closed without merging**